components/MapView.js

```diff
@@ -445,9 +445,79 @@ class MapView extends React.Component {
     this._runCommand('fitToCoordinates', [coordinates, edgePadding, animated]);
   }
 
-  takeSnapshot(width, height, region, callback) {
-    const finalRegion = region || this.props.region || this.props.initialRegion;
-    this._runCommand('takeSnapshot', [width, height, finalRegion, callback]);
+  /**
+   * Takes a snapshot of the map and saves it to a picture
+   * file or returns the image as a base64 encoded string.
+   *
+   * @param config Configuration options
+   * @param [config.width] Width of the rendered map-view (when omitted actual view width is used).
+   * @param [config.height] Height of the rendered map-view (when omitted actual height is used).
+   * @param [config.region] Region to render (Only supported on iOS).
+   * @param [config.format] Encoding format ('png', 'jpg') (default: 'png').
+   * @param [config.quality] Compression quality (only used for jpg) (default: 1.0).
+   * @param [config.result] Result format ('file', 'base64') (default: 'file').
+   *
+   * @return Promise Promise with either the file-uri or base64 encoded string
+   */
+  takeSnapshot(args) {
+    // For the time being we support the legacy API on iOS.
+    // This will be removed in a future release and only the
+    // new Promise style API shall be supported.
+    if (Platform.OS === 'ios' && (arguments.length === 4)) {
+      console.warn('Old takeSnapshot API has been deprecated; will be removed in the near future'); //eslint-disable-line
+      const width = arguments[0]; // eslint-disable-line
+      const height = arguments[1]; // eslint-disable-line
+      const region = arguments[2]; // eslint-disable-line
+      const callback = arguments[3]; // eslint-disable-line
+      this._runCommand('takeSnapshot', [
+        width || 0,
+        height || 0,
+        region || {},
+        'png',
+        1,
+        'legacy',
+        callback,
+      ]);
+      return undefined;
+    }
+
+    // Sanitize inputs
+    const config = {
+      width: args.width || 0,
+      height: args.height || 0,
+      region: args.region || {},
+      format: args.format || 'png',
+      quality: args.quality || 1.0,
+      result: args.result || 'file',
+    };
+    if ((config.format !== 'png') &&
```

Comment:
Consider pulling these out to constants at the top of the file, like:

```js
const VALID_FORMATS = ['png', 'jpg'];
const VALID_RESULTS = ['file', 'base64'];

// ...

if (!VALID_FORMATS.include(config.format)) {
  throw new Error('Invalid format specified');
}
```

Also it may be nice to use the [`invariant`](https://www.npmjs.com/package/invariant) package like:

```js
invariant(VALID_FORMATS.include(config.format), 'Invalid format specified');
```